src/DataTypes/MagicArray.php

```diff
@@ -35,6 +35,29 @@ public function __construct($value = null)
     }
 
     /**
+     *
+     * In PHP >=5.4.0, json_decode() accepts an options parameter, that allows you
+     * to specify that large ints should be treated as strings, rather than the
+     * PHP default behaviour of converting them to floats.
+     *
+     * Not all servers will support that, however, so for older versions we must
+     * manually detect large ints in the JSON string and quote them (thus converting
+     * them to strings) before decoding, hence the preg_replace() call.
+     *
+     * @link http://stackoverflow.com/a/27909889/1255333
+     * @param $input
+     */
+    private function json_decode_bigint_as_string($json, $assoc = false, $depth = 512) {
+        if (version_compare(PHP_VERSION, '5.4.0', '>=') && !(defined('JSON_C_VERSION') && PHP_INT_SIZE > 4)) {
+            return json_decode($json, $assoc, $depth);
+        } else {
+            $max_int_length = strlen((string) PHP_INT_MAX) - 1;
+            $json_without_bigints = preg_replace('/:\s*(-?\d{'.$max_int_length.',})/', ': "$1"', $json);
+            return json_decode($json_without_bigints, $assoc);
```

Comment:
Thanks for making the `$assoc` change :) I've just been thinking that we should probably also add $depth to the fallback json_encode?